### PR TITLE
[connector/routing] Supports matching the statement only once.

### PR DESCRIPTION
**Pull request by huange7**
**Description:** <Describe what has changed.>

Adding a feature: routing connector supports matching the statement only once

**Link to tracking Issue:** <Issue number if applicable> #26353

**Testing:** <Describe what testing was performed and which tests were added.>
Basic tests included for the functionality.

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Review by songy23 — COMMENTED** [review]
Please fix linter: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/6753227844/job/18405547771?pr=28888

File: connector/routingconnector/config.go
Comment (crobert-1) [inline connector/routingconnector/config.go:46]: Can you add the new config options to the README? Please include defaults (if applicable), usage description, and config examples.

**Review by crobert-1 — CHANGES_REQUESTED** [review]

File: connector/routingconnector/config.go
Comment (huange7) [reply connector/routingconnector/config.go:46]: Sure, I have added the new config options to the README. Please take a look and let me know if everything is clear and meets your expectations. Thank you!

File: connector/routingconnector/router.go
Comment (crobert-1) [inline connector/routingconnector/router.go:35]: I'm not very familiar with this connector, but in what scenario do we end up with multiple `table` items? (I may or may not have some functionality suggestions depending on the answer.)

**Review by crobert-1 — CHANGES_REQUESTED** [review]

File: connector/routingconnector/router.go
Comment (huange7) [reply connector/routingconnector/router.go:35]: Hi, @crobert-1 . Do you mean why the table is an array? Or why do we need `table`, `routes`, and `routeSlice` to store statements? If it's the former, it's because the `table` is used to contain multiple statements. If it's the latter, `routeSlice` is used to sort the statements, while `routes` is not used.
If I have misunderstood, please feel free to correct me.

File: connector/routingconnector/config.go
Comment (huange7) [reply connector/routingconnector/config.go:87]: You're absolutely right! In this way, we don't need to allow users to customize the order, but instead, we can directly adopt the order in which they are defined in the array. Besides, we need to explain this point in the README.

File: connector/routingconnector/router.go
Comment (djaglowski) [inline connector/routingconnector/router.go:75]: We should not need this. Ordered items are naturally managed in a slice.

**Review by djaglowski — CHANGES_REQUESTED** [review]
Please remove `routingItem.order` field.

**Comment by huange7** [comment]
> Please remove `routingItem.order` field.

done

File: connector/routingconnector/README.md
Comment (crobert-1) [inline connector/routingconnector/README.md:39]:
```suggestion
- `match_once (optional)`: determines whether the connector matches multiple statements or not. If enabled, the payload will be routed to the first pipeline in the `table` whose routing condition is met.
```
Just a nit, feel free to ignore. Just thought it was a little more succinct written this way.

Comment (crobert-1) [inline connector/routingconnector/README.md:39]:
```suggestion
- `match_once (optional, default: false)`: determines whether the connector matches multiple statements or not. If enabled, the payload will be routed to the pipeline whose routing condition is met, and is defined first in the `table`.
```
It would probably be helpful to include the default here.

File: connector/routingconnector/config.go
Comment (crobert-1) [inline connector/routingconnector/config.go:45]: Can we also set `MatchOnce` to `false` by default in the `factory.go:createDefaultConfig()` method? I think it would be best to be explicit.

File: connector/routingconnector/traces_test.go
Comment (crobert-1) [inline connector/routingconnector/traces_test.go:221]: Good test cases, thanks for adding! Another might be where it doesn't match the first, but does successfully match the second statement and gets routed there. If we update the values slightly it should be pretty simple to add another test.

Same comment goes for logs and metrics tests. We need to confirm that telemetry can be routed to the non-default and not-first `table` defined in case there's some kind of edge case we're missing.

File: connector/routingconnector/config.go
Comment (djaglowski) [reply connector/routingconnector/config.go:45]: I disagree here. Booleans have false as their zero value. We typically do not explicitly declare zero values for common data types.

Comment (crobert-1) [reply connector/routingconnector/config.go:45]: That's fair, especially if existing common practice here is to not explicitly set it.

File: connector/routingconnector/traces_test.go
Comment (huange7) [reply connector/routingconnector/traces_test.go:221]: > Good test cases, thanks for adding! Another might be where it doesn't match the first, but does successfully match the second statement and gets routed there. If we update the values slightly it should be pretty simple to add another test.

I think the test case "span matched by one expression, multiple pipelines" meets your requirements. It sets attributes["value"] to 5, which causes it to not match the first two rules but match the third rule "route() where attributes["value"] == 5". As a result, it sends the data to both tracesDefault and traces0.

File: connector/routingconnector/README.md
Comment (huange7) [reply connector/routingconnector/README.md:39]: Your expression is more concise.

File: connector/routingconnector/traces_test.go
Comment (crobert-1) [reply connector/routingconnector/traces_test.go:221]: Good point. I didn't put it together that it was sending to both `tracesDefault` and `traces0`, which will only happen in the case where that last condition is successfully met. 👍

**Review by crobert-1 — APPROVED** [review]

**Comment by djaglowski** [comment]
Related question which can be a separate issue. Is the _other_ (match multiple routes) approach implemented correctly? I am wondering if we are not properly fanning out the data.

For example, if we match routes 1 and 2, we cannot send the same copy of the data to both consumers. We need to use a fanout. In other words, we should build a consumer which contains all the pipelines in both 1 and 2, and send to that.

**Comment by huange7** [comment]
The PR hasn't been merged yet. Is there anything else I need to do?

**Comment by djaglowski** [comment]
I've opened a new issue to track my suggestions related to matching multiple routes.

@crobert-1, would you mind reviewing the open conversations and resolving as appropriate.

@huange7, anything else you need from us here other than another review?

**Review by djaglowski — APPROVED** [review]

**Comment by huange7** [comment]
> @huange7, anything else you need from us here other than another review?

Thank you very much for your support. I don't have any other questions.

**Review by crobert-1 — APPROVED** [review]

**Comment by crobert-1** [comment]
All of my concerns have been addressed, but I don't have permission to resolve the open conversations. This may be because I don't have [write access](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#resolving-conversations) in contrib, but maybe I'm missing something.

**Comment by huange7** [comment]
> @huange7 I think we're good to go as soon as the conflict is resolved and CI passes

Thanks. @djaglowski

**Pull request merged**
